app/src/main/java/com/nutomic/syncthingandroid/model/Completion.java

```diff
@@ -0,0 +1,138 @@
+package com.nutomic.syncthingandroid.model;
+
+import android.util.Log;
+
+import com.nutomic.syncthingandroid.util.DefaultHashMap;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+
+/**
+ * This class caches remote folder and device synchronization
+ * completion indicators defined in {@link CompletionInfo#CompletionInfo}
+ * according to syncthing's REST "/completion" JSON result schema.
+ * Completion model of syncthing's web UI is completion[deviceId][folderId]
+ */
+public class Completion {
+
+    private static final String TAG = "Completion";
+
+    DefaultHashMap<String, DefaultHashMap<String, CompletionInfo>> deviceFolderMap =
+        new DefaultHashMap<String, DefaultHashMap<String, CompletionInfo>>();
+
+    DefaultHashMap<String, CompletionInfo> defaultDevice = new DefaultHashMap<String, CompletionInfo>();
+
+    /**
+     * Removes a folder from the cache model.
+     */
+    private void removeFolder(String folderId) {
+        for (String deviceId : deviceFolderMap.keySet()) {
+            if (deviceFolderMap.get(deviceId).containsKey(folderId)) {
+                deviceFolderMap.get(deviceId).remove(folderId);
+                break;
+            }
+        }
+    }
+
+    /**
+     * Updates device and folder information in the cache model
+     * after a config update.
+     */
+    public void updateFromConfig(List<Device> newDevices, List<Folder> newFolders) {
+        // Handle devices that were removed from the config.
+        List<String> removedDevices = new ArrayList<>();;
+        Boolean deviceFound;
+        for (String deviceId : deviceFolderMap.keySet()) {
+            deviceFound = false;
+            for (Device device : newDevices) {
+                if (device.deviceID.equals(deviceId)) {
+                    deviceFound = true;
+                    break;
+                }
+            }
+            if (!deviceFound) {
+                removedDevices.add(deviceId);
+            }
+        }
+        for (String deviceId : removedDevices) {
+            Log.v(TAG, "updateFromConfig: Remove device '" + deviceId + "' from cache model");
+            deviceFolderMap.remove(deviceId);
+        }
+
+        // Handle devices that were added to the config.
+        for (Device device : newDevices) {
+            if (!deviceFolderMap.containsKey(device.deviceID)) {
+                Log.v(TAG, "updateFromConfig: Add device '" + device.deviceID + "' to cache model");
+                deviceFolderMap.put(device.deviceID, new DefaultHashMap<String, CompletionInfo>());
+            }
+        }
+
+        // Handle folders that were removed from the config.
+        List<String> removedFolders = new ArrayList<>();;
+        Boolean folderFound;
+        for (String deviceId : deviceFolderMap.keySet()) {

```

Comment:
You can just iterate over values here, as deviceId is not used anywhere other than getting the values.

Reply:
Yeah, and make it longer and import a new util.?!

Reply:
What util import? Just iterate over map.entrySet()

Reply:
Or even map.values()

Reply:
Entryset is a method on the map, I am suprised that needs an iterator, but if it does, let it be, makes the code more concise and potentially avoids the repeated gets.

Reply:
Hmm.. doesn't compile:
/home/builduser/syncthing-android/app/src/main/java/com/nutomic/syncthingandroid/model/Completion.java:26: error: incompatible types: Entry<String,HashMap<String,CompletionInfo>> cannot be converted to HashMap<String,HashMap<String,CompletionInfo>>
        for (HashMap<String, HashMap<String, CompletionInfo>> folderMap : deviceFolderMap.entrySet()) {
                                                                                                  ^

Reply:
Ok figured it out, import java.util.Map was required for entrySet.